### PR TITLE
script to change the database and add partitions

### PR DESCRIPTION
**Pull request by acn-dgopa**
## Related Issue(s)
- #{issue number}

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: test/Altinn.Auth.AuditLog.Tests/appsettings.test.json [hunk 1]
```diff
@@ -1,5 +1,10 @@
 {
   "PostgreSQLSettings": {
-    "EnableDBConnection": "false"
+    "EnableDBConnection": "true",
+    "WorkspacePath": "Migration",
+    "AdminConnectionString": "Host=localhost;Port=5432;Username=auth_auditlog_admin;Password={0};Database=authauditlogdb",
+    "ConnectionString": "Host=localhost;Port=5432;Username=auth_auditlog_admin;Password={0};Database=authauditlogdb",
```
Comment (Alxandr) [inline test/Altinn.Auth.AuditLog.Tests/appsettings.test.json:6]: This should not be admin.

File: src/Altinn.Auth.AuditLog.Core/Models/Partition.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Altinn.Auth.AuditLog.Core.Models
+{
+    /// <summary>
+    /// Used for partition creation
+    /// </summary>
+    public class Partition
+    {
+        public required string Name { get; set; }
+        public DateOnly StartDate { get; set; }
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog.Core/Models/Partition.cs:15]: These should probably also be required.

File: src/Altinn.Auth.AuditLog.Core/Models/Partition.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Altinn.Auth.AuditLog.Core.Models
+{
+    /// <summary>
+    /// Used for partition creation
+    /// </summary>
+    public class Partition
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog.Core/Models/Partition.cs:12]: Make this a `sealed record` - gives you a better `ToString()` for debugging, and it makes sense for this class to be immutable.

File: src/Altinn.Auth.AuditLog.Persistence/Altinn.Auth.AuditLog.Persistence.csproj [hunk 4]
```diff
@@ -4,14 +4,31 @@
...
+  <ItemGroup>
+    <InternalsVisibleTo Include="$(AssemblyName).Tests" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <EmbeddedResource Include="Migration/**/*.sql" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <None Remove="migration\v0.02\00-default-privileges.sql" />
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog.Persistence/Altinn.Auth.AuditLog.Persistence.csproj:32]: This should not be here.

File: src/Altinn.Auth.AuditLog.Persistence/AuthenticationEventRepository.cs [hunk 5]
```diff
@@ -18,25 +10,25 @@ namespace Altinn.Auth.AuditLog.Persistence
...
             _logger = logger;       
         }
 
+        /// <inheritdoc/>
         public async Task InsertAuthenticationEvent(AuthenticationEvent authenticationEvent)
         {
             const string INSERTAUTHNEVENT = /*strpsql*/@"
-            INSERT INTO authentication.eventlog(
+            INSERT INTO authentication.eventlogv1(
 	        sessionid,
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog.Persistence/AuthenticationEventRepository.cs:32]: Indentation issue - also, use `"""` strings for multiline queries.

File: src/Altinn.Auth.AuditLog.Persistence/AuthorizationEventRepository.cs [hunk 6]
```diff
@@ -38,7 +28,7 @@ public AuthorizationEventRepository(
         public async Task InsertAuthorizationEvent(AuthorizationEvent authorizationEvent)
         {
             const string INSERTAUTHZEVENT = /*strpsql*/@"
-            INSERT INTO authz.eventlog(
+            INSERT INTO authz.eventlogv1(
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog.Persistence/AuthorizationEventRepository.cs:31]: Indentation issue - also, use `"""` strings for multiline queries.

File: src/Altinn.Auth.AuditLog.Persistence/PartitionManagerRepository.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                        var cmd = batch.CreateBatchCommand();
+                        cmd.CommandText = /*strpsql*/$"""
+                            CREATE TABLE IF NOT EXISTS {partition.SchemaName}.{partition.Name}
+                            PARTITION OF {partition.SchemaName}.eventlogv1
+                            FOR VALUES FROM ('{partition.StartDate:yyyy-MM-dd}') TO ('{partition.EndDate:yyyy-MM-dd}')
+                            """;
+                        batch.BatchCommands.Add(cmd);
+                    }
+
+                    var result = await batch.ExecuteNonQueryAsync(cancellationToken);
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog.Persistence/PartitionManagerRepository.cs:91]: Is this correct? Do you get 0 if no partitions are created and non-zero if not?

File: src/Altinn.Auth.AuditLog/Services/PartitionCreationHostedService.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            // Create linked token to allow cancelling executing task from provided token
+            _stoppingCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+            _timer = _timeProvider.CreateTimer(
+                static (state) =>
+                {
+                    var self = (PartitionCreationHostedService)state!;
+                    self.CreateMonthlyPartitionFromTimer();
+                },
+                dueTime: TimeSpan.FromDays(1), // if this runs at once, change to 1 day
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog/Services/PartitionCreationHostedService.cs:49]: This comment should be removed.

File: src/Altinn.Auth.AuditLog/appsettings.AT22.json [hunk 9]
```diff
@@ -1,5 +1,6 @@
 {
   "KeyVaultSettings": {
     "SecretUri": "https://altinn-at22-auditlog-kv.vault.azure.net/"
-  }
-}
\ No newline at end of file
+  },
+  "kvSettings:SecretUri": "https://altinn-at22-auditlog-kv.vault.azure.net/"
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog/appsettings.AT22.json:5]: You need to do the same for all environments

File: src/Altinn.Auth.AuditLog/appsettings.Development.json [hunk 10]
```diff
@@ -4,5 +4,14 @@
       "Default": "Information",
       "Microsoft.AspNetCore": "Warning"
     }
+  },
+  "PostgreSQLSettings": {
```
Comment (Alxandr) [inline src/Altinn.Auth.AuditLog/appsettings.Development.json:8]: Use the newer settings format, `Altinn:Npgsql:auditlog`

File: test/Altinn.Auth.AuditLog.Tests/Altinn.Auth.AuditLog.Tests.csproj [hunk 11]
```diff
@@ -9,10 +9,17 @@
     <IsTestProject>true</IsTestProject>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Remove="Controllers\AuthenticationEventControllerTest - Copy.cs" />
```
Comment (Alxandr) [inline test/Altinn.Auth.AuditLog.Tests/Altinn.Auth.AuditLog.Tests.csproj:13]: This is definitely not correct.

File: test/Altinn.Auth.AuditLog.Tests/Controllers/AuthenticationEventControllerTest.cs [hunk 12]
```diff
@@ -19,35 +20,24 @@ namespace Altinn.Auth.AuditLog.Tests.Controllers
...
 
         [Fact]
         public async Task CreateAuthenticationEvent_Ok()
         {
+            using var client = CreateEventClient();
             AuthenticationEvent authenticationEvent = new AuthenticationEvent()
-            {                
-                Created = new DateTimeOffset(2018, 05, 15, 02, 05, 00, TimeSpan.Zero),
+            {
+                Created = TimeProvider.GetUtcNow(), //new DateTimeOffset(DateTime.Now.Year, 05, 15, 02, 05, 00, TimeSpan.Zero),
```
Comment (Alxandr) [inline test/Altinn.Auth.AuditLog.Tests/Controllers/AuthenticationEventControllerTest.cs:40]: Remove comment

File: test/Altinn.Auth.AuditLog.Tests/Utils/AdvanceableTimeProvider.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
+using System;
+
+namespace Altinn.Platform.Authentication.Tests.Utils;
+
+public class AdvanceableTimeProvider : TimeProvider
```
[the remark this answers is not in the document]
Comment (Alxandr) [reply test/Altinn.Auth.AuditLog.Tests/Utils/AdvanceableTimeProvider.cs:5]: This whole class can be deleted.

File: test/Altinn.Auth.AuditLog.Tests/appsettings.test.json [hunk 14]
```diff
@@ -1,5 +1,10 @@
 {
   "PostgreSQLSettings": {
```
Comment (Alxandr) [inline test/Altinn.Auth.AuditLog.Tests/appsettings.test.json:2]: All of this can be removed

**Review by Alxandr — APPROVED** [review]

File: src/Altinn.Auth.AuditLog.Persistence/PartitionManagerRepository.cs (hunk 7 above)
Comment (acn-dgopa) [reply src/Altinn.Auth.AuditLog.Persistence/PartitionManagerRepository.cs]: no, this always returns a -1. So what is the better way to validate the batch execution

**Pull request merged**